### ISSUE TITLE
Provider "enableReplaceCRD" not working

### ISSUE THREAD

**Issue by ncsibra** [pulumi/pulumi-kubernetes#1841]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

Related original issue: https://github.com/pulumi/pulumi-kubernetes/issues/1293
It was fixed in https://github.com/pulumi/pulumi-kubernetes/pull/1819 by @lblackstone .

I tested with the [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack), upgraded from `18.0.7` to `23.3.2`.
It looks like the `enableReplaceCRD` provider option does not replace the CRD as intended, at least on already existing resources.  
I set the flag, run a `pulumi up` and it does not show any update, I assume the provider should be updated.
Then updated the chart version and it failed on `prometheuses.monitoring.coreos.com` CRD.

```bash
Previewing update (dev)

View Live: https://app.pulumi.com/prmrgt-pulumi/k8s-prometheus-operator-crd/dev/previews/57b34987-feee-48dc-9431-d91bf36ab84c

     Type                                                               Name                                       Plan       Info
     pulumi:pulumi:Stack                                                k8s-prometheus-operator-crd-dev
     └─ kubernetes:helm.sh/v3:Chart                                     prometheus-operator
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  prometheusrules.monitoring.coreos.com      update     [diff: ~metadata]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  podmonitors.monitoring.coreos.com          update     [diff: ~metadata]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  servicemonitors.monitoring.coreos.com      update     [diff: ~metadata]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  probes.monitoring.coreos.com               update     [diff: ~metadata,spec]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  alertmanagerconfigs.monitoring.coreos.com  update     [diff: ~metadata,spec]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  alertmanagers.monitoring.coreos.com        update     [diff: ~metadata,spec]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  thanosrulers.monitoring.coreos.com         update     [diff: ~metadata,spec]
 ~      └─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  prometheuses.monitoring.coreos.com         update     [diff: ~metadata,spec]

Resources:
    ~ 8 to update
    3 unchanged

Do you want to perform this update? yes
Updating (dev)

View Live: https://app.pulumi.com/prmrgt-pulumi/k8s-prometheus-operator-crd/dev/updates/8

     Type                                                               Name                                       Status                  Info
     pulumi:pulumi:Stack                                                k8s-prometheus-operator-crd-dev            **failed**              1 error
     └─ kubernetes:helm.sh/v3:Chart                                     prometheus-operator
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  prometheusrules.monitoring.coreos.com      updated                 [diff: ~metadata]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  podmonitors.monitoring.coreos.com          updated                 [diff: ~metadata]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  servicemonitors.monitoring.coreos.com      updated                 [diff: ~metadata]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  probes.monitoring.coreos.com               updated                 [diff: ~metadata,spec]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  alertmanagerconfigs.monitoring.coreos.com  updated                 [diff: ~metadata,spec]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  alertmanagers.monitoring.coreos.com        updated                 [diff: ~metadata,spec]
 ~      ├─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  thanosrulers.monitoring.coreos.com         updated                 [diff: ~metadata,spec]
 ~      └─ kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition  prometheuses.monitoring.coreos.com         **updating failed**     [diff: ~metadata,spec]; 1 error

Diagnostics:
  pulumi:pulumi:Stack (k8s-prometheus-operator-crd-dev):
    error: update failed

  kubernetes:apiextensions.k8s.io/v1:CustomResourceDefinition (prometheuses.monitoring.coreos.com):
    error: 1 error occurred:
        * the Kubernetes API server reported that "prometheuses.monitoring.coreos.com" failed to fully initialize or become live: CustomResourceDefinition.apiextensions.k8s.io "prometheuses.monitoring.coreos.com" is invalid: metadata.annotations: Too long: must have at most 262144 bytes

Resources:
    ~ 7 updated
    3 unchanged

Duration: 23s
```

### Steps to reproduce

The whole code only deploy the CRD-s from the chart, other resources deployed in separate stack.

```typescript
import * as k8s from '@pulumi/kubernetes'

async function run(): Promise<void> {
  const k8sProvider = new k8s.Provider(`k8s-provider`, {
    enableReplaceCRD: true,
  })

  const namespace = 'monitoring'

  const values = {
    nameOverride: 'prometheus',
    fullnameOverride: 'prometheus',
    namespaceOverride: namespace,
    defaultRules: {
      create: false,
    },
    additionalPrometheusRules: [],
    global: {
      rbac: {
        create: false,
      },
      imagePullSecrets: [],
    },
    alertmanager: {
      enabled: false,
    },
    grafana: {
      enabled: false,
    },
    kubeApiServer: {
      enabled: false,
    },
    kubelet: {
      enabled: false,
    },
    kubeControllerManager: {
      enabled: false,
    },
    coreDns: {
      enabled: false,
    },
    kubeDns: {
      enabled: false,
    },
    kubeEtcd: {
      enabled: false,
    },
    kubeProxy: {
      enabled: false,
    },
    kubeScheduler: {
      enabled: false,
    },
    kubeStateMetrics: {
      enabled: false,
    },
    'kube-state-metrics': {
      rbac: {
        create: false,
      },
      podSecurityPolicy: {
        enabled: false,
      },
    },
    nodeExporter: {
      enabled: false,
    },
    'prometheus-node-exporter': {},
    prometheusOperator: {
      enabled: false,
      admissionWebhooks: {
        enabled: false,
      },
    },
    prometheus: {
      enabled: false,
    },
  }

  new k8s.helm.v3.Chart(
    'prometheus-operator',
    {
      repo: 'prometheus-community',
      chart: 'kube-prometheus-stack',
      version: '23.3.2',
      namespace: namespace,
      values: values,
    },
    {
      provider: k8sProvider,
    }
  )
}

run()

```

1. Deploy the chart with version `18.0.7`, without `enableReplaceCRD`.
2. Add `enableReplaceCRD` flag, then run `pulumi up`.
3. Update the chart version to `23.3.2`, then run `pulumi up`.

Expected: Update successfully.
Actual: Fails on `prometheuses.monitoring.coreos.com` with `the Kubernetes API server reported that "prometheuses.monitoring.coreos.com" failed to fully initialize or become live: CustomResourceDefinition.apiextensions.k8s.io "prometheuses.monitoring.coreos.com" is invalid: metadata.annotations: Too long: must have at most 262144 bytes`

Updating the CRD with `kubectl replace -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/v0.52.0/example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml` works without any issue, so the `enableReplaceCRD` does not work at all or does not do the same thing as `kubectl replace`.

Versions:
```bash
➜  pulumi (prometheus) ✗ pulumi version
v3.19.0
➜  pulumi (prometheus) ✗ yarn list --pattern '@pulumi'
yarn list v1.22.10
warning package.json: No license field
warning prmrgt-infra@0.0.0: No license field
├─ @pulumi/aws@4.32.0
├─ @pulumi/cloudflare@4.0.0
├─ @pulumi/docker@3.1.0
├─ @pulumi/kubernetes@3.12.1
├─ @pulumi/pulumi@3.19.0
├─ @pulumi/query@0.3.0
├─ @pulumi/random@4.3.1
├─ @pulumi/tls@4.1.0
└─ @pulumi/vault@4.6.0
Done in 0.15s.
➜  pulumi (prometheus) ✗ pulumi plugin ls
NAME        KIND      VERSION  SIZE    INSTALLED  LAST USED
aws         resource  4.32.0   355 MB  n/a        1 hour ago
aws         resource  4.24.1   286 MB  n/a        2 months ago
aws         resource  4.21.1   281 MB  n/a        2 months ago
aws         resource  4.19.0   279 MB  n/a        2 months ago
aws         resource  4.10.0   270 MB  n/a        2 months ago
aws         resource  4.0.0    256 MB  n/a        2 months ago
aws         resource  3.33.0   250 MB  n/a        2 months ago
aws         resource  3.22.0   259 MB  n/a        2 months ago
aws         resource  3.19.2   255 MB  n/a        2 months ago
aws         resource  3.5.0    244 MB  n/a        2 months ago
aws         resource  3.4.0    243 MB  n/a        2 months ago
aws         resource  3.1.0    235 MB  n/a        2 months ago
aws         resource  2.13.1   228 MB  n/a        2 months ago
aws         resource  2.13.0   228 MB  n/a        2 months ago
aws         resource  2.11.0   227 MB  n/a        2 months ago
aws         resource  2.9.1    222 MB  n/a        2 months ago
aws         resource  2.7.0    220 MB  n/a        2 months ago
aws         resource  2.6.1    220 MB  n/a        2 months ago
aws         resource  1.29.0   228 MB  n/a        2 months ago
aws         resource  1.20.0   218 MB  n/a        2 months ago
aws         resource  1.18.0   200 MB  n/a        2 months ago
aws         resource  1.13.0   192 MB  n/a        2 months ago
aws         resource  1.7.0    221 MB  n/a        2 months ago
cloudflare  resource  4.0.0    43 MB   n/a        1 hour ago
cloudflare  resource  3.5.0    43 MB   n/a        2 months ago
cloudflare  resource  3.3.0    42 MB   n/a        2 months ago
cloudflare  resource  3.2.0    42 MB   n/a        2 months ago
cloudflare  resource  2.14.0   41 MB   n/a        2 months ago
cloudflare  resource  2.10.0   46 MB   n/a        2 months ago
cloudflare  resource  2.9.0    46 MB   n/a        2 months ago
cloudflare  resource  2.6.0    47 MB   n/a        2 months ago
cloudflare  resource  2.4.0    39 MB   n/a        2 months ago
cloudflare  resource  2.3.1    39 MB   n/a        2 months ago
cloudflare  resource  1.9.0    38 MB   n/a        2 months ago
cloudflare  resource  1.6.0    38 MB   n/a        2 months ago
cloudflare  resource  1.2.0    43 MB   n/a        2 months ago
docker      resource  3.1.0    40 MB   n/a        2 months ago
docker      resource  3.0.0    41 MB   n/a        2 months ago
kubernetes  resource  3.12.1   77 MB   n/a        1 hour ago
kubernetes  resource  3.8.1    77 MB   n/a        2 months ago
kubernetes  resource  3.7.2    76 MB   n/a        2 months ago
kubernetes  resource  3.7.0    76 MB   n/a        2 months ago
kubernetes  resource  3.5.0    67 MB   n/a        2 months ago
kubernetes  resource  2.8.2    77 MB   n/a        2 months ago
kubernetes  resource  2.7.5    75 MB   n/a        2 months ago
kubernetes  resource  2.7.4    75 MB   n/a        2 months ago
kubernetes  resource  2.6.1    80 MB   n/a        2 months ago
kubernetes  resource  2.4.3    70 MB   n/a        2 months ago
kubernetes  resource  2.4.2    70 MB   n/a        2 months ago
kubernetes  resource  2.3.0    53 MB   n/a        2 months ago
kubernetes  resource  2.2.2    53 MB   n/a        2 months ago
kubernetes  resource  1.6.0    50 MB   n/a        2 months ago
kubernetes  resource  1.5.6    50 MB   n/a        2 months ago
random      resource  4.3.1    32 MB   n/a        2 months ago
random      resource  4.2.0    31 MB   n/a        2 months ago
random      resource  3.1.0    30 MB   n/a        2 months ago
random      resource  3.0.0    31 MB   n/a        2 months ago
random      resource  2.4.4    40 MB   n/a        2 months ago
random      resource  2.3.1    43 MB   n/a        2 months ago
random      resource  2.2.0    39 MB   n/a        2 months ago
random      resource  2.1.2    37 MB   n/a        2 months ago
tls         resource  4.1.0    32 MB   n/a        1 hour ago
tls         resource  4.0.0    31 MB   n/a        2 months ago
tls         resource  3.3.0    30 MB   n/a        2 months ago
tls         resource  3.0.0    31 MB   n/a        2 months ago
tls         resource  2.4.2    40 MB   n/a        2 months ago
tls         resource  2.3.0    43 MB   n/a        2 months ago
tls         resource  2.2.0    39 MB   n/a        2 months ago
vault       resource  4.6.0    46 MB   n/a        1 hour ago
vault       resource  4.5.1    44 MB   n/a        2 months ago
vault       resource  4.4.0    44 MB   n/a        2 months ago
vault       resource  4.1.0    44 MB   n/a        2 months ago
vault       resource  3.5.0    42 MB   n/a        2 months ago
vault       resource  3.2.0    49 MB   n/a        2 months ago
vault       resource  3.1.0    49 MB   n/a        2 months ago
vault       resource  2.4.0    48 MB   n/a        2 months ago
vault       resource  2.3.0    48 MB   n/a        2 months ago
vault       resource  2.2.1    46 MB   n/a        2 months ago

TOTAL plugin cache size: 8.3 GB
```

**Comment by lblackstone** [comment]
Based on the error message, it looks like you're running into a [different issue](https://github.com/pulumi/pulumi-kubernetes/issues/1048) on the update:
`metadata.annotations: Too long: must have at most 262144 bytes`

Can you remove that annotation and try again?

**Comment by ncsibra** [comment]
Removing the annotation does not help, getting the same error.

I'm aware of that issue, but if `enableReplaceCRD` is really equivalent to `kubectl replace`, then this shouldn't be a problem.
As I wrote in the first comment, `kubectl replace` works without any issue.
It basically removes the `kubectl.kubernetes.io/last-applied-configuration` annotation and does not add it again after replace.

So I assumed, that `enableReplaceCRD` should solve https://github.com/pulumi/pulumi-kubernetes/issues/1048 too, at least for CRDs.

**Comment by ncsibra** [comment]
During testing, I noticed another strange behavior, if I should put it another issue, let me know.
Looks like pulumi falsely states, that the `prometheuses.monitoring.coreos.com` CRD was updated when `enableReplaceCRD` is false, but actually, it wasn't.
I tested this on minikube with kubernetes version `v1.21.2`.

Steps to reproduce:
1. Make sure you're using an older pulumi-kubernetes version than `3.12.1`, I used `3.7.2`.
2. Use the code from my first comment with `enableReplaceCRD: false` and deploy the chart version `18.0.7`. 
    It should be deployed fine and the `prometheuses.monitoring.coreos.com` CRD should have a long `kubectl.kubernetes.io/last-applied-configuration` annotation.
3. Save the CRD definition with `kubectl get crd/prometheuses.monitoring.coreos.com -o yaml`.
4. Update the chart to `23.3.2` and try to deploy, it should fail on `prometheuses.monitoring.coreos.com` with the message `the Kubernetes API server reported that "prometheuses.monitoring.coreos.com" failed to fully initialize or become live: CustomResourceDefinition.apiextensions.k8s.io "prometheuses.monitoring.coreos.com" is invalid: metadata.annotations: Too long: must have at most 262144 bytes`.
5. Update pulumi-kubernetes to `3.12.1`.
6. Try to deploy again, it will show, that all CRDs were updated.
7. Save the CRD definition again and compare it with the previous version, it will be the same.
8. Optionally check pulumi state with `pulumi stack export`, it will show that `prometheuses.monitoring.coreos.com` CRD resource has the newest version, both in the `inputs` and `outputs` section.
   The easiest way to do this, just to check the `controller-gen.kubebuilder.io/version` annotation, for the `18.0.7` chart it will be `v0.4.1`, for `23.3.2` will show `v0.6.2`.
9. Set `enableReplaceCRD: true` then run `pulumi up`.
10. Run `pulumi refresh`, it will show, that `~__inputs` changed on `prometheuses.monitoring.coreos.com`.
11. Optionally check pulumi state again, now both `inputs` and `outputs` will show the older CRD definition from the `18.0.7` chart.
12. Run `pulumi up`, `prometheuses.monitoring.coreos.com` will show an update, but will fail with the error: `the Kubernetes API server reported that "prometheuses.monitoring.coreos.com" failed to fully initialize or become live: CustomResourceDefinition.apiextensions.k8s.io "prometheuses.monitoring.coreos.com" is invalid: metadata.annotations: Too long: must have at most 262144 bytes`.

So with the latest pulumi-kubernetes version without `enableReplaceCRD`, it says that every CRD updated, even `prometheuses.monitoring.coreos.com`, but on the environment, it actually stays the same.

**Comment by lblackstone** [comment]
Alright, thanks for the update, and for your patience on this. CRDs have been a particularly tricky area to handle, so the detailed information is really helpful in figuring out the problems. I added this issue to our milestone, so we'll give it another look shortly.

> if I should put it another issue, let me know

Let's leave it here for now, and we can split it out if it turns out to be a different underlying issue.

**Comment by tamirhad** [comment]
@ncsibra do you have a workaround? i cant install Prometheus :(
@lblackstone  where can i track on that issue in the meantime?

**Comment by ncsibra** [comment]
@lblackstone The original issue is solved with version 3.14.1.
But the other issue, described in [this comment](https://github.com/pulumi/pulumi-kubernetes/issues/1841#issuecomment-996760823) still exists.
With `enableReplaceCRD: false`, pulumi still shows that `prometheuses.monitoring.coreos.com` was updated, but on the environment, it stays the same, and after a refresh, pulumi realizes this too.

**Comment by lblackstone** [comment]
@ncsibra Whoops, sorry for missing that! I opened a separate issue and PR with a fix.
